# # Purpose:  Bioinformatics and Computational Biology project for BCB330
# #
# #
# # Version:  1.0
# #
# # Date:     2018 03
# # Author:   Judy Heewon Lee (heewon.lee@mail.utoronto.ca)
# #
# # Versions:
# #
# #           1.0 First implementation of AQE
# #
# #
# # ============================= AQE Sample Script ============================ #
#
# #TOC> ==========================================================================
# #TOC>
# #TOC>   Section  Title                                                    Line
# #TOC> --------------------------------------------------------------------------
# #TOC>   1        Using the Database                                        33
# #TOC>   2        Fetching a Sequence                                       96
# #TOC>   3        Writing a Sequence                                       125
# #TOC>   4        Fetching an Alignment                                    141
# #TOC>   5        Writing an Alignment                                     154
# #TOC>   6        Fetching an Alignment Generated by ClustalW,             169
# #TOC>            ClustalOmega or Muscle
# #TOC>   7        Writing an Alignment Generated by ClustalW,              192
# #TOC>            ClustalOmega or Muscle
# #TOC>   8        Scoring Alignments                                       213
# #TOC>
# #TOC> ==========================================================================
#
# # ======================= 1. Using the Database ============================ #
#
# # The database is generated by the function createDB() in AQE. It generates
# # the referenceDB which is provided as an .rda file in the package. It must
# # be noted that the reference database must be labelled as referenceDB
# # and be contained in the global env. in order for many functions in AQE
# # to work. You can create the database using the following command:
#
# referenceDB <- createDB()
#
# # referenceDB is list categorized into 4 sections:
# #
# # 1. $category
# # 2. $alnCategory
# # 3. $alignments
# # 4. $seqCategory
# #
#
# # The $category is a data frame which contains the different reference
# # categories in BaliBASE and the corresponding descriptions about them.
# # You can access this by using the notation below:
#
# category <- referenceDB$category
# (category)
#
# # The $alnCategory is a data frame consisting of all the alignment ID's from
# # BaliBASE and the corresponding reference category they belong to. You can
# # use the function getAlnID by filling the referenceCat parameter.
#
# alnList <- getAlnID(referenceCat = "RV50")
# (alnList)
#
# # The $alignments category is a list of alignment ID's containing the matrix
# # form of the alignment. The alignments can be accessed using the notation:
#
# aln <- referenceDB$alignments[["BB11001"]]
# (aln)
#
# # Or, you may simply use the function fetchAln which retrieves the alignment
# # with different attributes. See the Fetch Alignment section for more
# # details.
#
# # The last category in the referenceDB database is the $seqCategory
# # which is a data frame that contains the sequence ID and the
# # corresponding alignment ID's. The seqCategory is mainly used
# # as a 'view'. It is a searchable table that will make it
# # efficient to access values in the other categories such as $alignments.
#
# # The seqID's to a specific alignment can be retrieved through this function:
#
# sequences <- getSeqID("BB11004")
# (sequences)
#
# # You can also retrieve all sequences that are under a broader category.
#
# sequences <- getSeqID("BB11")
#
# # The alnID's that contain a specific sequence ID can be retrieved through
# # this function:
#
# alnIDs <- getAlnID(seqID = "1ton_")
# (alnIDs)
#
# # ======================= 2. Fetching a sequence ============================ #
#
# # Assume we want to find the sequence for the specific gene '1aab_'.
# # We can simply use fetchSeq in order to retrieve it.
#
# mySeq <- fetchSeq("1aab_")
# (mySeq)
#
# # [1] "GKGDPKKPRGKMSSYAFFVQTSREEHKKKHPDASVNFSEFSKKCSERWKTMSAKEKGKF
# # EDMAKADKARYEREMKTYIPPKGE"
#
# # This time, we would like to fetch the sequence without collapsing it and
# # in matrix format. We can simply change the collapse and asMatrix
# # parameters to do so.
#
# mySeq <- fetchSeq("1aab_", collapse = FALSE, asMatrix = TRUE)
# (mySeq)
#
# #[[1]]
# # [1] "-" "-" "-" "G" "K" "G" "D" "P" "K" "K" "P" "R" "G" "K" "M" "S"
# # "S" "Y" "A" "F" "F" "V" "Q" "T" "S" "R" "E"
# # [28] "E" "H" "K" "K" "K" "H" "P" "D" "A" "S" "V" "N" "F" "S" "E" "F"
# # "S" "K" "K" "C" "S" "E" "R" "W" "K" "T" "M"
# # [55] "S" "A" "K" "E" "K" "G" "K" "F" "E" "D" "M" "A" "K" "A" "D" "K"
# # "A" "R" "Y" "E" "R" "E" "M" "K" "T" "Y" "I"
# # [82] "P" "P" "K" "G" "E" "-" "-" "-" "-" "-" "-" "-" "-" "-" "-"
#
# # The output above is shifted for formalities.
#
# # ======================= 3. Writing a sequence ============================= #
#
# # We can also write the sequence of interest in .fasta format into a
# # given directory. Written sequence files will end up in 'data/Output'
# # if no directories are given.
# # Let's assume we want to write the sequence '1aab_' in a file
#
# writeSeq("1aab_")
#
# # The file 1aab_.fasta will be produced in 'data/Output'. We can also
# # change the name of the file and place it in a different directory.
#
# writeSeq("1aab_", fileName = "Hello", directory = "data")
#
# # A file named Hello.fasta should be located in the data folder.
#
# # ====================== 4. Fetching an alignment =========================== #
#
# # Similar to fetching sequences, we can retrieve alignments from BaliBASE
# # as well. Let's say we wanted the alignment "BB20025" from BaliBASE,
# # non-collapsed and in matrix form.
#
# myAln <- fetchAln("BB20025", collapse = FALSE)
#
# # asMatrix and collapse are automatically set to TRUE.
# # We can also use the list of alignment ID's to retrieve the alignment from
# # referenceDB using fetchAln. Refer to the Using the Database section for
# # more information.
#
# # ======================= 5. Writing an alignment =========================== #
#
# # The function writeAln will write an alignment for the user in multifasta
# # format.
# # For example, if we wanted the alignment "BB11001" and wanted the file to be
# # named "test1". We also want to collapse the alignment (remove gaps).
# # We can then run the following:
#
# writeAln("BB11001", fileName = "test1", collapse = TRUE)
#
# # There should be a file in "data/Output" named "test1.mfa" containing the
# # collapsed sequences of alignment BB11001.
# # You can also change the directory of the alignment file using the
# # directory parameter.
#
# # ==== 6. Fetching an alignment from ClustalW, ClustalOmega or Muscle ======= #
#
# # A functionality included in AQE is fetching alignments from ClustalW,
# # ClustalOmega and Muscle using the msa package. Let's assume we
# # want to align the sequences from "BB12042" alignment using Muscle.
#
# msa <- fetchMuscle("BB12042", readType = "AA")
# (msa)
#
# # We can also control certain attributes in the alignment. Let's assume
# # we want to change the gap extension penalty to 1 instead of keeping it
# # to its default value, 0.
#
# msa <- fetchMuscle("BB12042", readType = "AA", gapExtension = 1)
# (msa)
#
# # If we wanted to use our own .fasta files, we can also input the
# # filepath to the .mfa file containing collapsed sequences in
# # fasta format. Example:
#
# msa <- fetchClustalOmega("data/Output/test1.mfa", readType = "AA")
# (msa)
#
# # ==== 7. Writing an alignment from ClustalW, ClustalOmega or Muscle ======== #
#
# # Another function that is added to AQE is to write the alignments produced
# # by ClustalW, ClustalOmega and Muscle. Let's assume we want an alignment
# # generated by ClustalW using our test1.mfa file made using writeAln.
#
# writeClustalW("data/Output/test1.mfa", readType = "AA")
#
# # There should be a file called "test1ClustalW.mfa" in the "data/Output"
# # directory containing the aligned sequences from test1 in multifasta
# # format.
#
# # The user can also provide an alignment ID from the BaliBASE database
# # to align the sequences in the BaliBASE alignment using the alignment
# # algorithms.
#
# writeClustalOmega("BB40048", readType = "AA", fileName = "myAln")
#
# # The user can also change the file destination by providing the
# # directory parameter.
#
# # ======================= 8. Scoring alignments =========================== #
#
# # Scoring alignments is another function included in AQE
# # We can return details on alignment comparisons using the function
# # scoreAlignments. Let's assume we want to compare a ClustalW alignment
# # with the alignments provided by BaliBASE. We will choose alignment
# # "BB11001" to do this.
#
# writeAln("BB11001", collapse = FALSE, fileName = "reference")
# writeClustalW("BB11001", readType = "AA", fileName = "testAln")
#
# pac <- scoreAlignment("data/Output/reference.mfa",
#                       "data/Output/testAln.mfa", SP = TRUE,
#                       CS = TRUE, details = TRUE)
# (pac)
#
# # Setting SP and CS as TRUE returns the properties of the alignments
# # based on the scoring. Setting Details as true returns a pairwise
# # alignment comparison (PAC) object from the AlignStat package. PAC is
# # subsettable, so it is easy to find specific attributes of interest
# # from the object. For more details, visit their documentation:
# # https://cran.r-project.org/web/packages/AlignStat/AlignStat.pdf
#
# # If we simply want the scores of the two alignments, we can do so
# # by setting details to FALSE.
#
# scores <- scoreAlignment("data/Output/reference.mfa",
#                          "data/Output/testAln.mfa", SP = TRUE,
#                          CS = TRUE, details = FALSE)
# (scores)
#
# # scoreAlignment will then return a data frame containing the
# # corresponding scores.
#
# #                 Type     Score
# # 1 Total Column Score 0.7291667
# # 2 Sum of Pairs Score 0.9341564
#
#
#
#
#
#
